### ISSUE TITLE
Split JSON settings editor discussion

### ISSUE THREAD

**Issue by roblourens** [microsoft/vscode#125952]
I'd like to start a discussion around simplifying the split JSON settings editor to make it easier for us to maintain.

Keeping "deprecated" features around for a long time is not free and leads to a trail of tech debt that costs developer time that could be spent improving other areas of VS Code. It also represents many lines of code, and as the complexity and download size of VS Code grows, we need to take available opportunities to reduce it. I know it still has some fans, but telemetry shows that it is used very little compared to other features.

I want to think about how to keep some of the benefits of the split JSON editor implemented using typical VS Code UX patterns, rather than custom code just for one editor. Reading through past discussion on the settings editor, it's clear that the most valued element of the split editor is the ability to read the default settings and the user's overrides easily on one screen. We can actually do this easily in VS Code by showing the default settings as a simple text editor side by side with the overrides using the same sort of split editor that we use for diffing.

![image](https://user-images.githubusercontent.com/323878/121579382-84a20500-c9e0-11eb-86c2-3b74f71c5243.png)

This lets us clean up code related to the custom search bar, the scope switcher, the pencil icon for editing settings, and some other UI decorations. We could even have an action to toggle the default settings split on or off, which would make this feature useful to people who currently just use the basic json editor. I think this plan is a good balance between serving this user workflow and software maintenance cost.

The settings editor GUI remains for fuzzy search and switching scopes quickly.

**Comment by eamodio** [comment]
While I will miss the fuzzy search filtering, I think this is a good compromise. As for switching scopes, could we maybe just have a toggle button in the editor toolbar to switch from user to workspace settings?

**Comment by eamodio** [comment]
Not sure I'm following exactly -- typing `settings ` would list the available scopes and then open either the GUI or JSON based on your settings? I still think a quick toggle in the editor toolbar would be helpful -- but maybe the button just shows a menu with those same choices (`User`, `Workspace`, etc)

**Comment by roblourens** [comment]
Just JSON. GUI has it's own scope switcher

**Comment by jeanp413** [comment]
Oh no! I've been using the split JSON editor since the beginning.
Hope there's more feedback from the community to reconsider this decision if possible :disappointed:

**Comment by jeanp413** [comment]
@eamodio Definitely the search box to filter the settings and the pencil action to quickly add it to my settings.json instead of typing or copying some setting

**Comment by kkm000** [comment]
The fuzzy search box is priceless. That would be a huge loss of productivity for me. I can copy stuff between panes manually, but it seems to me that finding _that_ setting without fuzzy search would not be simply possible. Now it's so intuitive that I am finding what I want to change by trying 2-3 fuzzy searches, and sometimes hit the bullseye from the first attempt. Please keep it if at all possible. Pretty please.

It's like Emacs Ido mode, only much better. Please don't kill it.

**Comment by carlocardella** [comment]
I second @kkm000 about the fuzzy search. Using the GUI for fussy search, copy the setting id and then move to the json settings editor seems to overcomplicate things, I see lots of back and forth between panels and frustration in my future if the fuzzy search in the json editor does away...

**Comment by NatoBoram** [comment]
There's many places where the split JSON settings could be improved while it's changed and these changes could probably be reflected in the rest of the editor, without being tied to settings specifically.

* The fuzzy search should be added as an option to the default search box.
* Searching in one panel should optionally search on the other side, too
* Selecting a key either side should focus on that key on the other panel
  * That's similar to the Markdown behaviour, but it could be used with any _key-value_-based languages like JSON/YAML/Properties

These improvements could work in many split-panel scenarios. For example, editing i18n files in a project could use this kind of behaviour.

**Comment by wenfangdu** [comment]
I'd say the settings JSON split editor is one of the best features of VSCode, using it from the start, tried the UI version, never a fan of it.

**Comment by kandros** [comment]
This settings page was the first feature that really pushed me to give vscode a try, is so natural to use and easy to share settings

**Comment by OneOfOne** [comment]
Please leave the json settings editor, the GUI one is annoying to use.

**Comment by ryan3E0** [comment]
The split JSON editor makes it easy to see exactly what is overridden in the user settings, workspace, and folder at a glance. The features being cited to removed (custom search bar, the scope switcher, the pencil icon for editing settings) streamline the experience. JSON should go back to the default, in my opinion. Telemetry is showing not a lot of people using it because the default was changed to the UI version, that doesn't make it superior. I'd also guess that those who went in and changed it may have also disabled telemetry.

I'm not too knowledgeable in the code for these features but what are we talking about in terms of actually maintaining these? It's only tech debt if it is costly (paying "interest" on that debt). As you said, it is custom code just for one editor, can't it exist and function without any need to constantly update it? It is a JSON editor with features on top, which doesn't seem like it would be in flux all the time. If it were intertwined with a bunch of other features, yeah I could see that being a headache and a huge slowdown.

I'm not sure I'm buying the download size either, how much larger are we talking? Surely there are much bigger fish to fry when it comes to the app size.

**Comment by zaatas** [comment]
Should this occur, I'll leave VScode!!  I will take the dive and make the investment into Neovim and Lua.

**Comment by pcjmfranken** [comment]
If there are only resources to streamline and maintain either the visual menu or the JSON editor, can we have a pol which one to keep? And to save some time, work on removing the visual menu can start right away :)

But in all seriousness, to me, some of the JSON view's main benefits are:

- Higher information density: Quickly see what settings and options are available
- Discover "hidden"/undocumented options in-app
- Very clear overview of your setting overrides
- The ability to annotate your overrides thanks to the `jsonc` format
- Overrides 
- An easy way to copy settings and overrides for use or documentation elsewhere
- Less clicking, more clacking

The visual editor will never come close to the JSON one in those areas. I do not believe that the minor seeming benefits of the proposed changes outweigh the negative impact those will have on the things that make the JSON editor so fantastic!

**Comment by Mellbourn** [comment]
I would really miss the ability to comment my settings. I often write explanations for me to remember motives for specific settings.

**Comment by JohnSpindler** [comment]
> I would really miss the ability to comment my settings. I often write explanations for me to remember motives for specific settings.

+1 to this in addition to adding in some clarifying notes for myself when a setting isn't documented in a way I understand.

**Comment by ZUIcat** [comment]
I don't like GUI, I would rather remove GUI instead of Json.

**Comment by roblourens** [comment]
I agree that the fuzzy search is useful, and it would be interesting to experiment with a way to search settings using this same search logic using a quick pick control. This seems like a way to tackle it that works within existing VS Code UI conventions, without changing the meaning of the basic editor find widget, or other special logic. It's hard for me to think about how to apply the fuzzy search (or pencil icon) to generic JSON documents since it's implementation is very settings-specific, and I don't think there are other real examples of this split pane defaults/overrides editor pattern.

@ryan3E0 It feels like it should be a standalone thing that we never have to touch, but it exists as a component within a larger system that is often in flux. For example, as we've made changes to how opening an editor works in VS Code to support custom editors and notebooks, we recently had to spend a lot of developer hours changing the split JSON editor code to adapt to these changes. And we periodically make cross-cutting changes across the entire codebase or across feature areas which can impact this code. 

Just in case this point gets lost in the thread, I want to be clear, the thing that is being changed is only the custom split-pane JSON editor. Editing settings in a JSON file will never go away.

**Comment by carlocardella** [comment]
@roblourens, I think the _quality_ of the JSON settings editing experience is important. I understand the JSON settings file will not go away (a similar point was discussed the last time this topic was brought previously: #62964) and I do not have a reason to doubt you guys are motivated to do what is best, but I also understand why people may be wary of this change. 

For comparison, Atom (I just installed it after a really long time to check) has eerily similar Settings and Keybindings GUIs and offers the ability to edit the config file directly (it's yaml in Atom's case). Anyway there is no advanced editing experience, there is no intellisense, I have not found a way to have a side-by-side view with both default settings and my settings (which as other mentioned, I found invaluable to learn about settings and their use, and see at a glance what is the default and what I am using as override). In other words, it is not really usable (one might as well just open the file in Notepad).

So, I understand this will change (I'm not sure how much room for discussion is left since this issue is already scheduled for the July release), I really hope the experience will not degrade too much. It would be a real shame.

You know what? I would be really curious to see the reaction if you opened a similar issue to remove the Settings GUI, I'd be curioius to see how many people would miss it :smile:

**Comment by kkm000** [comment]
@roblourens, I hear you, and I think that I understand this feeling. I have been a software engineer for most of my working life, and I know that some parts of code look like a burden. I experienced this too many times: a feature is a mess, and it just does not fit the rest of design, but users like it. This happened both in open-source and for-bread-and-butter projects. You see how many people came here to say how they find the feature indispensable. And sometimes it takes more effort to keep it alive.

Think of this. I have read Raymond Chen's blog for years, and his book too, and can relate to engineers having to jump through hoops on fire to keep the Windows features that even make no sense—like keeping it backward compatible with bugs in enterprise software whose developers had been long out of business. That's necessary for Windows business, but not really a gratifying part of the job. And here you have two dozens of flesh-and-blood ~~GitHub nicks~~ fellow engineers begging to keep the feature. Yes, that's a minuscule number compared to the total number of VSCode users, but we're not just a number, we're the real hardcore ones who push the pedal to the metal. And there are much more who did not notice the announce, or were too busy to voice their displeasure. Please think of all of us who you can make happy by keeping it, or an equally efficient equivalent.

Settings is always a tough topic to discuss; a complex highly customizable environment—take Visual Studio, VSCode, Emacs—have thousands of them. For example, key bindings in Visual Studio have a search box, but it first appeared in VS2012, IIRC. How tortuous was it to simply change a key binding back then. VSCode beats the other two in this small sample in its simplicity of finding and changing what you want to customize: Emacs has M-x apropos to search both customize variables and descriptions, not fuzzy; Visual Studio does not have anything even close for settings; neither comes close to a fuzzy search in a single pane which reflects the changes as you type. 

Then, for VSCode it is _much more essential_ than for Emacs. For the latter, there is Emacs.SO where I can go and find or ask about a particular customization. VSCode is very agile, and in my experience the _vast majority_ of customizing recipes that I'm finding on the internet do not work any more when they are a year or two old, without minor or major tweaks. Options search is perhaps the only way for advanced users to keep up with the changes. It's not a "nice to have" feature. It's an essential feature, really.

**Comment by jeanp413** [comment]
@roblourens, I didn't know this is the third time (after #62964 and #64932) bringing up this subject, somehow I missed those previous discussions and surely many other people missed them and will miss this one too.
So, if you are looking for reasons to **not** simplify the split JSON editor, you should revisit those previous discussions where the stated arguments against doing it are in concord with the ones stated here.

**Comment by carlocardella** [comment]
@roblourens would it make code maintenance easier if the split JSON settings editor as we see it today was a built-in extension instead of part of the main product? 

I know _logically_ a settings editor should be part of the main product but if _functionally_ for users the JSON settings editor can remain as it is today, if it will simplify the main VSCode codebase and the split JSON settings editor can be maintained and receive future improvements, then it may be a good compromise to consider?

**Comment by thernstig** [comment]
I would be a bit vary of this statistic:

> telemetry shows that it is used very little compared to other features

That might very well be true looking at the numbers only. But if we recognize that the GUI editor is the _default_ and people are not aware of the split JSON editor, and they had an option to select either or when installing VS Code, it would be a much more fair comparison.

Statistics is what it is, you can prove any point if you don't take a holistic approach.

**Comment by ladyrick** [comment]
https://github.com/microsoft/vscode/issues/62964 has 49 👎 .

https://github.com/microsoft/vscode/issues/64932 has 19 👎 .

And this issue has 29 👎  for now.

Why are you so confident to delete the split JSON settings editor?

Split JSON settings editor is not a tech debt. It's a great feature of VSCode. It's the best way to handle large number of config items.

Your effort spent to maintain this feature is valuable. Please don't remote it.

**Comment by carlocardella** [comment]
> telemetry shows that it is used very little compared to other features
>> That might very well be true looking at the numbers only. But if we recognize that the GUI editor is the default and people are not aware of the split JSON editor, and they had an option to select either or when installing VS Code, it would be a much more fair comparison.

Good point @thernstig. I would add that comparing the settings editor "to other features" may be misleading, it depends on which _other features_ one uses for comparison. I assume people do not need to change settings all the time, the pure number of "settings update" events will of course be relatively low. 
Again, I think it would be interesting to compare the JSON settings editor telemetry with the GUI editor, keeping in mind that the GUI version is the default, therefore guaranteed to have higher usage for the reasons @thernstig mentioned (and more, I'm sure).

**Comment by lijhj** [comment]
I just got 57 settings so far. I edit them and sort them by alphabet manually. I copy and paste the json content when I'm on a new machine. 

I like these 2 best. They make the text very easy to read:
    "editor.fontSize": 14,
    "editor.fontWeight": "bold",

It's better to make json the default and kill the ui option:
"workbench.settings.editor": "json", //ui,

{
    "breadcrumbs.enabled": false,
    "C_Cpp.errorSquiggles": "Disabled",
    "C_Cpp.formatting": "Disabled",
    "C_Cpp.updateChannel": "Insiders",
    "debug.showBreakpointsInOverviewRuler": true,
    "diffEditor.renderSideBySide": false,
    "editor.autoClosingBrackets": "never",
    "editor.autoClosingQuotes": "never",
    "editor.dragAndDrop": false,
    "editor.folding": false,
    "editor.fontSize": 14,
    "editor.fontWeight": "bold",
    "editor.hover.enabled": false,
    // "editor.lineNumbers": "off", //
    "editor.minimap.enabled": false,
    "editor.renderLineHighlight": "all",
    "editor.roundedSelection": false,
    "editor.rulers": [80], //
    // "editor.wordWrap": "wordWrapColumn", //wordwrap
    "explorer.confirmDelete": false,
    "explorer.confirmDragAndDrop": false,
    "explorer.openEditors.sortOrder": "alphabetical",
    "extensions.ignoreRecommendations": true,
    "extensions.showRecommendationsOnlyOnDemand": true,
    "files.exclude": {
        "**/.build":true,
        "**/CVS": true,
        "**/._.DS_Store": true,
        "**/.DS_Store": true,
        "**/.git": true,
        "**/.gitignore": true,
        "**/.hg": true,
        "**/.svn": true,
        "**/.swiftpm":true,
        "**/.vscode": true,
        "**/*.d": true, //dep files
        "**/*.o": true, //obj files
        "**/*.qbk": true,
    },
    "files.insertFinalNewline": true,
    "files.trimTrailingWhitespace": true,
    "git.autofetch": true,
    "git.enableSmartCommit": true,
    "http.proxySupport": "off",
    "outline.showArrays": false,
    "outline.showBooleans": false,
    "outline.showConstants": false,
    "outline.showEnums": false,
    "outline.showFields": false,
    "outline.showVariables": false,
    "window.title": "${activeEditorLong}",
    "workbench.colorTheme": "Default Dark+",
    "workbench.editor.untitled.hint": "hidden",
    "workbench.settings.editor": "json", //ui, json
    "workbench.settings.useSplitJSON": true,
    "workbench.startupEditor": "newUntitledFile",
    "files.autoSave": "afterDelay",
}

**Comment by roblourens** [comment]
I think that making the comparison to Windows' efforts for backcompat is useful in that it illustrates how different the two products are in what our goals and values are, our development resources, and what people like about them. Do people say they appreciate that vscode will jump through hoops to maintain every feature for years, or that it is lightweight, agile, and focused on the essentials? 😁 

Hadn't thought about it much, but I think it would be possible for an extension to implement a settings editor. Porting something like the split json editor to a webview editor extension would be a substantial amount of work, would require bundling an editor in the extension, and isn't something that we would do. The URIs for the settings schemas are here: https://github.com/microsoft/vscode/blob/37a11814295fe9e90542d8990c30b83bb2de42fd/src/vs/workbench/services/configuration/common/configuration.ts#L18

**Comment by pcjmfranken** [comment]
> jump through hoops to maintain every feature for years, or that it is lightweight, agile

Nope, you're absolutely right! We're starting to get through to you I see 😉

An impractical GUI editor that will always be lagging behind core features/API developments on a functional level, but then also has a bunch of UI components that aren't really used anywhere else in the entire editor that require constant upkeep. Doesn't sound very lightweight and agile to me, more like a high overhead costly blocking factor.

A JSON editor on the other hand.... I mean, here, from the [official language description](https://www.json.org/json-en.html):
>JSON (JavaScript Object Notation) is a lightweight data-interchange format. It is easy for humans to read and write. It is easy for machines to parse and generate.

**Comment by roblourens** [comment]
@pcjmfranken Just in case it's not clear, the thing that is being changed is only the custom split-pane JSON editor. It seems like the points you've brought up are mainly relevant to editing JSON in a basic text editor as well. Editing settings in a JSON file will never go away.

**Comment by pcjmfranken** [comment]
@roblourens A fair point that's easy to forget and good to be reminded of.

I guess, how I personally see it, is that the functionality that you are aiming to remove is essentially the cherry on top of a [really awesome cream pie](https://lekkerboxmeer.nl/wp-content/uploads/2019/02/Slagroomtaart-rond-300x375.jpg), whilst telling us that your pecan pie isn't too bad either. I like the cream pie better though ([and who wouldn't, right?!](https://www.supervlaai.nl/vestigingen/webshop/producten/slagroomschnitte1601380884422.jpg)), and let's be honest it's just not the same without the cherry!

**Comment by lonix1** [comment]
Agree with all users here, we'd like the json config "experience" to stay. If changes must be made, then please at least ensure that the new editor has the same features/ux as the old one.

And while you're at it, please consider [fixing](https://github.com/microsoft/vscode/issues/114909) the [bugs](https://github.com/microsoft/vscode/issues/87764) in the current json editor:
- when clicking to another tab and then returning, the left pane's tree is collapsed (frustrating)
- the left pane's nodes are not sorted (very very frustrating!, as it's hard to find stuff)

Interested to see what you guys come up with. Thanks for involving us.

**Comment by angel3cu** [comment]
I would like to give a thumbs up for the classic JSON settings editor. I have NEVER actually used the UI editor, due to the large amount of settings available on VS Code. I currently have more than 1,000 different settings in total which is impossible to maintain using the UI. My actual JSON settings file is 274 lines long!

If I had to choose, I would keep the JSON editor instead of the clunky UI. Any improvements to it are welcome, but please do not remove it. 🙂

**Comment by carlocardella** [comment]
The fuzzy search bar is gone in today's Insiders. Now to search for settings one must use the GUI and then switch back to the JSON editor (while it is still around in this mutilated form).
Bad productivity and workflow impact 👎😫

**Comment by carlocardella** [comment]
The pencil is also gone. At least settings autocomplete still works.

Another side effect (this will likely have a more limited impact): I had a custom keybinding to search my settings (below), this no longer works since the _editor_ is no longer recognized as "SettingsEditor"; is looks like a _simple_ file where `ctrl+f` invokes the search widget

```JSON
{
    "key": "ctrl+shift+f",
    "command": "actions.find",
    "when": "inSettingsEditor"
}
```

**Comment by carlocardella** [comment]
I have disabled automatic updates in Stable, I think I'll stick with 1.58 till I absolutely have to upgrade

```json
"update.enableWindowsBackgroundUpdates": false
```

**Comment by OneOfOne** [comment]
Good ol' microsoft...

> majority of users like and use feature X
> microsoft, nah, add Y and make it the default
> majority of users ignore Y
> microsoft, well, i wasted time developing Y, so screw all those people, I'll just remove X and keep Y as the only option.

Oh well, at least it's open source, we all can just switch vscodium or neovim if it comes to it.

Also you're using telemetry as a metric, however did you factor in that the majority of users who know how to use the json editor will more than likely disable telemetry?

Please reconsider, there are so many other bugs that's been open for years that could use more love instead of deleting a feature that works.

**Comment by roblourens** [comment]
I've pushed a change to replace the split JSON editor with the simplified split JSON editor - just two panes with defaults and settings.

@carlocardella the pencil should still be there! I realized that we already used this in the basic settings.json file, so I've left it, contrary to what I said before.

**Comment by jeanp413** [comment]
> I've pushed a change to replace the split JSON editor with the simplified split JSON editor - just two panes with defaults and settings.

@roblourens so what was the point of creating this issue a calling it a "discussion" if you are gonna ignore us all and do it either way 892785264b4682d102b9cb81e01a549c218a74ee ? :-1:  

> the pencil should still be there!

Personally, the pencil is more useful when it's on the left pane which is now gone. Anyway, the most useful feature is the fuzzy search box which is gone too, now it feels like opening two notepad windows side by side.

**Comment by carlocardella** [comment]
> @carlocardella the pencil should still be there! I realized that we already used this in the basic settings.json file, so I've left it, contrary to what I said before.

@roblourens, no, the pencil on the left pane (which is where it was most useful) is gone

**Comment by carlocardella** [comment]
History repeats itself apparently: https://github.com/microsoft/vscode/issues/64932#issuecomment-447093051

I can only hope it will repeat completely and, as that December over 2 years ago, the decision is reversed. I will not upgrade beyond 1.58 unless absolutely necessary.

**Comment by pcjmfranken** [comment]
The fuzzy search box is probably the most useful feature. Very disappointed to see it go!

@jeanp413 makes a very good regarding telemetry data representativeness. It could very well be that users who prefer the JSON editor  are more likely to disable it. This is the case for 3/4 devs I discussed this with earlier today.

**Comment by wenfangdu** [comment]
The choice between 2399 lines of code or this feature is a no-brainer for me:

![image](https://user-images.githubusercontent.com/28700378/125874951-0fc8dd28-ff81-4175-94a3-326ca89702e6.png)

**Comment by roblourens** [comment]
>> the pencil should still be there!

> Personally, the pencil is more useful when it's on the left pane which is now gone.

Ah, yes, sorry.

I know this isn't much, but I was wondering whether it would help with search to be able to open a settings UI editor to the side. Then you'd be able to search there on the side while still keeping your focus on the setting.json text editor.

**Comment by carlocardella** [comment]
Something like this (here I'm just showing the Settings GUI and my actual JSON settings side-by-side, didn't have time to play with a more realisting mockup)? 

It looks like an ugly Frankenstein to me, an hack just to save a few clicks to the user to open the JSON file side by side like this, I do not see a real benefit. 
On the usability I have doubts but I would need to try it once first to be objective. 

![image](https://user-images.githubusercontent.com/5784415/125881709-61538a61-eacc-4a1e-b27a-eb6486768815.png)

**Comment by NatoBoram** [comment]
The JSON editor was gutted, but none of its features made their way to the text editor.

* There's no fuzzy search
* Searching for a setting doesn't hide unrelated settings
* Searching on one panel doesn't search in the other panel
* Selecting a key on either side does not scroll to that key on the other side
* The little pencil icon with the "copy this key/value to the other pane" is completely gone

All of these are improvements that should've been ported to the text editor before the settings editor was gutted.

I'm sad that you did not consider how the settings editor made VSCode the behemoth it is today. Without it, VSCode would be *nothing* today.

**Comment by carlocardella** [comment]
I do not know if VSCode would be _nothing_ without the JSON settings editor but it is definitely a huge part of why I like and use VSCode every day instead of Visual Studio, or Atom, or other editors. I suspect I am not alone.

VSCode allows to customize and fine tune _every little detail_, from appearance to how the editor behaves, how extensions behaves, keybindings, the ability to install it on a new machine and be up and running with all extensions and settings sync'ed in just a few clicks and a couple of minutes and so on; all this flexibility has no price for someone who spends _hours_ each day in the code editor.

VSCode is the first editor I really felt _mine_ and I spend time to try out and improve these customizations and I like to keep them well organized. Most of that would simply not be possible with the GUI editor. Jumping back and forth between the two would be even worse.

**Comment by kkm000** [comment]
@roblourens:

> I think that making the comparison to Windows' efforts for backcompat is useful in that it illustrates how different the two products are in what our goals and values are, our development resources, and what people like about them. Do people say they appreciate that vscode will jump through hoops to maintain every feature for years, or that it is lightweight, agile, and focused on the essentials? 😁

You first build a strawman out of my words, reading into them a comparison of efforts for backward compatibility between VSCode and Windows, which is just not there, and asserting that I said that **every** feature must be maintained, although neither I, nor, to the best of my knowledge, obviously didn't. Yes, of course, in your exposition I do indeed look like an idiot throwing nonsensical comparisons and irrational requests. Then you beat this strawman victoriously, then add that grin-in-your-face thing.  _And you're speaking for Microsoft._ Okay. I admit that my appeal to your presumed engineering professionalism and the sense of satisfaction with your work was a logical error.

I joined the Insiders to provide feedback and contribute to the product development, not to spend my only lifetime on participating in mock "discussions" only to be made a laughing stock of by a Microsoft employee. As @carlocardella said, 1.58 Stable is a good choice. With telemetry off, naturally: it's none of Microsoft's business. I'll do my best not to assist your employer in any way in future, as long as it condones the public speaking policy of its employees mocking its customers.

By the way, you introduced what looks to me like a bug into the editor (maybe it's another "enhancement", I have no idea). The open `settings.json` tab now has no path name. Not in the title bar, not in the tab tooltip. So I have open tabs with  `settings.json`,  `settings.json` and  `settings.json`, and am trying to figure out which is which. The only way to discover the path is through the Explorer Open Editors section, where the path can be copied via the Copy Path command then examined by pasting it somewhere. If that is intentional, please do not waste your time on mocking me again: I've signed off updates to this thread.

**Comment by hsimpson** [comment]
> I know it still has some fans, but telemetry shows that it is used very little compared to other features.

The reason for this is, because the professional users use the JSON split pane editor because in enterprise environments and multiple projects this is much more productive then the fancy GUI Editor.

But the majority of users are probably not enterprise driven or multi project users, or sometimes beginner users who stick with the default. So the telemetry data for this is uselses in my opinion.

**Comment by x80819091** [comment]
@roblourens  In fact, I think YOU should be removed from the VSCode team.

Let's make this the most downvoted issue on the VSCode repo!

**Comment by JLChamberlain** [comment]
I've been keeping my eye on this issue to see how it progresses but never bothered to voice my opinion due to the clear dislike of this idea and assumed that it would just be scrapped, I thought that the developers would actually listen to us users and take on board our reactions. I'm rather disappointed and shocked to see it's still going ahead despite the clear dislike of this idea and the overall negative reaction from users regarding this change. What's the point of having a discussion about this if the overall voice regarding this idea is basically ignored?

**Comment by daniele-orlando** [comment]
@roblourens why don't you extract current implementation, making an optional extension that user could download in case they want the legacy behaviour?

**Comment by wenfangdu** [comment]
The split JSON editor should go back to the default, perhaps it has a bit steeper learning curve than the UI editor, but ultimately, you'll be able to work much faster and more comfortably in the split JSON editor.

Unfortunately, VSCode is going backward.

**Comment by dsugiyama** [comment]
I'm afraid that many of the people commenting here are misunderstanding the situation. Even if the *split* JSON settings editor is removed, you can still use the *normal* JSON editor, which has been the default in VS Code for a long time now. You won't be forced to use the GUI editor.

What will be removed:
<img width="1001" alt="split-json-editor" src="https://user-images.githubusercontent.com/5922503/126107314-181a56d1-f648-4851-b448-fbe64493a93b.png">

What will **not** be removed:
<img width="591" alt="normal-json-editor" src="https://user-images.githubusercontent.com/5922503/126107337-2f320479-71c3-49a9-b34d-0cc3a97c4fa4.png">

**Comment by wenfangdu** [comment]
@dsugiyama With critical functions being removed from the split JSON editor, we **are** forced to use the GUI editor. Of course, the plain JSON editor won't be removed, otherwise, what VSCode is for?

**Comment by dsugiyama** [comment]
@wenfangdu I'm not denying the usefulness of the split JSON editor, and personally I'd like to see it remains if development resouce allows. I just wanted to clear up some people's misconception.

I think at least some alternatives (for example, porting features like fuzzy search to the plain editor) should be considered to avoid significantly degrading the settings editing experience.

**Comment by pcjmfranken** [comment]
> significantly degrading the settings editing experience

When your more amiable users start using these words to describe the post-change landscape with...

**Comment by carlocardella** [comment]
@dsugiyama except maybe a couple comments, in all others I do not see misconception. This topic has already been discussed over two years ago, it was clear back then and I think it is clear now. 

The features removed are exactly the ones that make the JSON settings editor superior. Editing a JSON file with stone intellisense is the base minimum experience one expects these days from a code editor, there is no magic there. 

The fuzzy search, ability to see all the defaults and quickly compare to my custom settings,find a setting even if I  do not know the exact name, keep it all well organized (and other "features" mentioned in the thread) is what really makes the difference, why for many the JSON settings editor is the superior (and default, at least in my case) experience. 

Sorry, the problem statement is clear and the intent is also clear (it already landed last week in Insiders). I just think it is based on wrong assumptions.

**Comment by joseph1125** [comment]
Would it be possible to keep the split GUI editor as an extension?

**Comment by wenfangdu** [comment]
@roblourens It is now the most downvoted open issue on the VSCode repro, can this change your mind?

![image](https://user-images.githubusercontent.com/28700378/126326320-2a518a49-5a49-4406-af09-09675b85e6ec.png)

**Comment by JulienGrv** [comment]
I would be fine seeing the JSON editor being removed if the GUI or plain text editor where suitable alternative. However, I can see for myself one important missing feature (maybe there is more): where can one see the default values in the GUI or the plain text editor?

The GUI has 'reset settings' but it's not the same as showing the default value. For the plain text editor, the default value will appear after using the autocompletion but not while typing which is not ideal.

So, for me at least, removing the JSON split editor IMO harms the user experience.

**Comment by refcell** [comment]
No!!! don't do this, please.

**Comment by lonix1** [comment]
Another reason why I like the old editor "experience", is that I can easily commit config to source control. I can view diffs, and I can reset changes made by naughty extensions without my permission. That config is used by all team members, and most/all projects.

For this use case, the old editor makes it very easy, whereas the gui editor does not. If the new editor will not have the full "experience" (fuzzy search, descriptions, user vs workspace, etc.) then for this use case it will be a step backward.

Please ensure the new editor will have the same features/ux as the old editor.

**Comment by tepinly** [comment]
I don't understand why this update is still going to happen even though nobody is content it?

Why not just add a toggle button for the simplified JSON version while keeping the old version all the same?

**Comment by andrewarchi** [comment]
I've been using VS Code since at least 2016 and have accumulated a 1012-line-long settings.json file. Fuzzy search is invaluable in discovering and comparing settings and the pencil is great for enumerating options. The split-pane JSON view is so effective compared to GUI-oriented configuration in other IDEs, which suffer from poor discoverability, low information density, and slow click-based UI.

I disable telemetry almost without exception, and I would guess that many JSON configuration users do the same. I've just enabled telemetry in the hopes that my preferences will be more statistically known.

Please don't nerf the JSON config.

**Comment by redomar** [comment]
My hands were forced to block any future updates, in json split view no less.
`"update.mode": "none"`

**Comment by NatoBoram** [comment]
I feel like this issue isn't much of a discussion more than a simple announcement that the maintainers don't actually care about their users. How about switching the perspective of this discussion?

* https://github.com/microsoft/vscode/issues/129594

Users already made their voices heard (#62964 🔒, #64932 🔒, #125952). I feel like something is missing; changing tone should grant some new insights.

**Comment by lijhj** [comment]
how about using quotation marks for: "exact search",
if no quotation marks then: fuzzy search.

this is what macOS preview does for PDF files.

**Comment by Cheffu** [comment]
I know different people mentioned these points already, but just out of my view: 
- The split JSON settings editor including the pencil icons, the split view, the (fuzzy) search bar, ... is the best settings UI ever. In fact, it was the main feature for me to switch to VSCode a few years ago. It is also the feature I used to convince colleagues and friends to use VSCode.
- I think that many people using the JSON settings editor are exactly the ones deactivating the telemetry. Thus "telemetry shows that it is used very little compared to other features" is a useless argument against this feature.
- If you really want to reduce tech debt regarding settings, remove the GUI.

**Comment by poshcodebear** [comment]
I think @NatoBoram makes an excellent point with their new discussion: every argument used about technical debt and maintainability can be directed at the GUI settings interface as well.

I only just today saw this was happening, probably because for the most part I'm not digging through my settings file regularly.  I've not had to make any changes to it over the last month, and with only days to spare, discovered that this discussion has been happening that will remove all of the features I've long come to rely on when I'm tweaking my settings.  And that pretty much all the feedback is being ignored (you say you're not ignoring it, but then turn around and move forward with doing what you want anyway in spite of not really a single person voicing support of this decision, and only 13 thumb-ups on it vs 85 thumb-downs as of this point in time).

You've made statements to support your decision (telemetry and technical debt) and when those are called out, you don't bother addressing them (well, technically, you did respond to the technical debt criticisms, though again, the same arguments could be used against the GUI editor, so not exactly a satisfactory response there either).

At this point, you've made a lot of assumptions based on telemetry (ignoring things like the Tyranny of the Default™ and the assumption that telemetry alone will paint a real picture when it can and frequently is turned off by the power users who will prefer the JSON split-editor and who are the ones most loudly telling people they should use VSCode).  You've had a lot of feedback (which will also probably be biased, in all honesty, because the only people likely to know this is happening found it like I did, when our settings editor of choice let us know it's going away and points us here to this "discussion").  I think the appropriate course of action at this point is to step back from making this change in the July release and spend more time actually trying to figure out if this is the right decision or not.

Until then, I'm also blocking updates.  While I don't hop into my settings every day, sometimes not even every month, the GUI is simply not an adequate replacement when I need to change settings (for the reasons already listed by others) and a basic JSON file is simply nowhere near as functional as the split-editor (again, for the reasons already listed by others).

Oh, and I loaded Insiders for the first time in a long while just to see how this change would work.  The settings JSON is completely broken for me, it won't even let me save the danged thing or use the pencil edit because the file is "dirty" and needs to be saved.  But it won't let me save it because "the contents of the file is newer".  So, you've managed to not only remove everything that made the JSON split-editor so powerful, you've somehow managed to make it even less useful than Notepad, since at least Notepad will let you save files.

Please listen to your power users.  Table this change.  The features you're talking about removing are worth keeping around.

**Comment by roblourens** [comment]
I've read the feedback and I know that this isn't the outcome you all want, sorry. Thanks to those of you who tried out the new simple split editor. I hope you all give the simple editor or the GUI a chance at some point. Lots of people have pointed out that power users using the split editor are more likely to have disabled telemetry, that may be true, what telemetry says is not the main motivating factor here so maybe I shouldn't have brought it up.

> The settings JSON is completely broken for me, it won't even let me save the danged thing or use the pencil edit because the file is "dirty" and needs to be saved. But it won't let me save it because "the contents of the file is newer".

It seems like the pencil icon doesn't work right when the settings.json file is dirty. That should be fixed but I see the same behavior in the split editor in 1.58. Is that what you're seeing? I filed https://github.com/microsoft/vscode/issues/129861 for this.

@kkm000
> The open settings.json tab now has no path name. Not in the title bar, not in the tab tooltip.

If you haven't disabled breadcrumbs, you will see the path below the tab when the editor is active, but I will improve this next month: https://github.com/microsoft/vscode/issues/129856

Also, if anyone would like to try to create an extension with a custom editor that implements a settings editor, this would not be easy but I think it should theoretically be possible. The settings schemas can be loaded from an extension via these URIs: https://github.com/microsoft/vscode/blob/37a11814295fe9e90542d8990c30b83bb2de42fd/src/vs/workbench/services/configuration/common/configuration.ts#L18

**Comment by NatoBoram** [comment]
Aight, I'd like to point out the irony behind the `feature-request` label on a feature removal that no user actually requested. I know it's just a procedure, but you know... It still hurts.

Since you weren't listening in the first place, then I guess you can lock the issue before the release of the public build so that people surprised by a sudden heavy downgrade of functionality can come here, confused and dismayed, and really *feel* the warm welcome of a locked thread?

**Comment by wenfangdu** [comment]
@NatoBoram I guess he thought after a long time since the GUI editor became the default settings editor, more people will support his decision, but we hardcore split JSON editor fans haven't died out yet and even increased given the downvote count. He can't see us die out in a near future, hence decided to do it anyway.

**Comment by carlocardella** [comment]
The split editor as it is today is useless to me, at this point it would make sense to just get rid of the entire thing and leave the simple JSON file, why bother with the split vire now? It's nothing more than two files side by side

**Comment by TheWyo** [comment]
Found my way here from the little note that's now appeared in the split JSON editor's view and this is possibly the most baffling 'discussion' I've ever seen.  
You've done the equivalent of walking up to someone holding a swiss army knife, taking it off them and handing them a corkscrew instead, and then when questioned about it, have reassured them they can still open wine bottles.  
They're well aware of that fact, but what about everything else they used the knife for?

**Comment by pcjmfranken** [comment]
In the future, @roblourens, I suggest you either find a more suitable word for... well, whatever this was, or, better yet, take an example from our friends in China, North Korea, or Belarus and don't bother with a farce like this at all.

As insignificant as the removal of this feature may be in the grand scheme, the way this is handled is nothing less than insulting and has left a very bad aftertaste. Goodwill is not an unlimited resource, you know.

**Comment by ionut-botizan** [comment]
For me, the split JSON editor doesn't make much sense these days. I've been using it from the beginning and it was great (at first). However, my user settings file has grown to **700+ lines of settings** (no comments, just settings) and finding anything in that file is a pain.
I'd rather have a regular JSON editor with IntelliSense and regular search and that would be enough.

Besides, how often does anyone go into the settings file anyway? You'll do it a few times in the beginning, when you're getting started with VSCode, then you'll just go occasionally, when a new feature is added or (maybe) when you install a new extension.

**Comment by carlocardella** [comment]
Everyone has his workflow; my settings file is over 1200 lines (I'm pretty sure many people in this thread have large settings files), I do not edit it every single day but fairly often. Often enough to make what we have today (the GUI or the now broken split editor) very inefficient to use

**Comment by vadimcn** [comment]
Those wishing to suspend updates on Ubuntu, run this: `sudo apt-mark hold code`

**Comment by KerkDovan** [comment]
The release notes don't even mention the removal of fuzzy search! People who rarely open settings.json but read the release notes won't even know about this change. The purpose of this is to prevent a wave of negativity immediately after the update? Otherwise, I see no point in such silence - this is not an insignificant change.

**Comment by wenfangdu** [comment]
> The release notes don't even mention the removal of fuzzy search! People who rarely open settings.json but read the release notes won't even know about this change. The purpose of this is to prevent a wave of negativity immediately after the update? Otherwise, I see no point in such silence - this is not an insignificant change.

I wish I could thumb down [this release note](https://github.com/microsoft/vscode/releases/tag/1.59.0), and there even isn't a way to do it.

To disable further updates, use:

```json
{
  "update.enableWindowsBackgroundUpdates": false,
  "update.mode": "manual"
}
```

**Comment by wenfangdu** [comment]
> The release notes don't even mention the removal of fuzzy search! People who rarely open settings.json but read the release notes won't even know about this change. The purpose of this is to prevent a wave of negativity immediately after the update? Otherwise, I see no point in such silence - this is not an insignificant change.

I wish I could thumb down [this release note](https://github.com/microsoft/vscode/releases/tag/1.59.0), and there's not even a way to do it.

**Comment by wenfangdu** [comment]
To disable further updates, use the following settings (ironically, you can't even do this in one step via the GUI editor):

```json
{
  "update.enableWindowsBackgroundUpdates": false,
  "update.mode": "manual"
}
```

**Comment by tifrel** [comment]
As many of the commenters, I came from a tiny little notification in my settings editor to read this abomination of a discussion. I don't know how many people might have missed the notification. And how many will be taken by surprise.

I am not sure if it's only me, but I used to avoid anything from Microsoft like it's the plague. VS Code was different, as it was just too good of a product to ignore. VS Code made me reconsider my stance on MS. I know it's a big brand, and mass compatibility counts more than power user friendliness. VS Code struck the balance, as it was easy to get started with for the masses, yet highly efficient/productive for power users. I know this balance is hard, and that is one of the reasons I liked VS Code so much. Mutilating this tool, and doing so despite all the negative, yet constructive feedback, is sticking the finger to the power users and early adopters. And thus, the better image that MS painted themselves over the past few years breaks down into shattered pieces. It doesn't matter now because most users will never read this thread or not even notice the change. But you leave room for others to attract the power users, as you deprecate their beloved features. If they find something better than VS Code, they will move, and they will draw the crowd with them.

Congratulations, you paid technical debt by taking on usability debt.

**Comment by carlocardella** [comment]
Here we go, it didn't take long, one day after Stable got updated: #130245

**Comment by carlocardella** [comment]
#130262

**Comment by lijhj** [comment]
> To disable further updates, use the following settings (ironically, you can't even do this in one step via the GUI editor):
> 

If you refuse to receive updates there's no reason to stay. Both Qt creator and Eclipse have outline / symbol list feature and are cross platform and worth a try.

**Comment by Ticmea** [comment]
Like many others I have visited this "discussion" from time to time to vote on the comments. I didn't comment myself because what I wanted to say was already said by other users and I didn't want to spam this issue. But now as this has reached stable I feel like I have to voice my opinion as a comment, since it feels like the votes on the comments are being ignored (given that this was implemented despite the overwhelming opposition to this change).

For now I will stick to 1.58.2 for a while and see if maybe our voice will be heard when stable users notice the change and (presumably) voice their opinion here.

**Comment by tifrel** [comment]
PSA: Links to last usable version
- [Source](https://github.com/microsoft/vscode/releases/tag/1.58.2)
- [Binary](https://github.com/VSCodium/vscodium/releases/tag/1.58.2) (VSCodium, compiled for several architectures and in several package formats)
If someone knows a good guide of how to build/install from source, please let me know.

**Comment by x80819091** [comment]
@roblourens Congrats! It is now [the most downvoted issue](https://github.com/microsoft/vscode/issues?q=is%3Aissue+is%3Aclosed+sort%3Areactions--1-desc) in the entire vscode repo, well done!

**Comment by OneOfOne** [comment]
Judging by the lack of response from Microsoft, this is pretty much over.

So for those who want the last stable version (official links):

https://update.code.visualstudio.com/1.58.2/linux-x64/stable => vscode_linux_x64_1.58.2.tar.gz
https://update.code.visualstudio.com/1.58.2/linux-rpm-x64/stable => vscode_linux_x64_1.58.2.rpm
https://update.code.visualstudio.com/1.58.2/linux-deb-x64/stable => vscode_linux_x64_1.58.2.deb
https://update.code.visualstudio.com/1.58.2/linux-arm64/stable => vscode_linux_arm64_1.58.2.tar.gz
https://update.code.visualstudio.com/1.58.2/linux-rpm-arm64/stable => vscode_linux_arm64_1.58.2.rpm
https://update.code.visualstudio.com/1.58.2/linux-deb-arm64/stable => vscode_linux_arm64_1.58.2.deb
https://update.code.visualstudio.com/1.58.2/linux-armhf/stable => vscode_linux_armhf_1.58.2.tar.gz
https://update.code.visualstudio.com/1.58.2/linux-rpm-armhf/stable => vscode_linux_armhf_1.58.2.rpm
https://update.code.visualstudio.com/1.58.2/linux-deb-armhf/stable => vscode_linux_armhf_1.58.2.deb
https://update.code.visualstudio.com/1.58.2/darwin/stable => vscode_darwin_1.58.2.zip
https://update.code.visualstudio.com/1.58.2/darwin-arm64/stable => vscode_darwin_arm64_1.58.2.zip
https://update.code.visualstudio.com/1.58.2/darwin-universal/stable => vscode_darwin_universal_1.58.2.zip
https://update.code.visualstudio.com/1.58.2/win32/stable => vscode_win32_1.58.2.exe
https://update.code.visualstudio.com/1.58.2/win32-x64/stable => vscode_win32_1.58.2.exe
https://update.code.visualstudio.com/1.58.2/win32-arm64/stable => vscode_win32_1.58.2.exe
https://update.code.visualstudio.com/1.58.2/win32-user/stable => vscode_win32_user_1.58.2.exe
https://update.code.visualstudio.com/1.58.2/win32-x64-user/stable => vscode_win32_user_1.58.2.exe
https://update.code.visualstudio.com/1.58.2/win32-arm64-user/stable => vscode_win32_user_1.58.2.exe
https://update.code.visualstudio.com/1.58.2/win32-archive/stable => vscode_win32_archive_1.58.2.zip
https://update.code.visualstudio.com/1.58.2/win32-x64-archive/stable => vscode_win32_archive_1.58.2.zip
https://update.code.visualstudio.com/1.58.2/win32-arm64-archive/stable => vscode_win32_archive_1.58.2.zip

Make sure to disable automatic updates.

I'm also maintaining a copy on my site at https://oneofone.dev/pkgs/vscode-1.58.2/ in case they decide to remove the files, feel free to use the links, or use the download script to host them yourself.

Time to re-learn vim or emacs I guess.

**Comment by x80819091** [comment]
Seriously, who're the leaders of this guy? Call them in. From what I've seen so far, a microsoft developer can do whatever he wants for the part he's assigned without any user agreement?

![image](https://user-images.githubusercontent.com/86513712/128620891-5ed0ca37-c987-49f8-9dcf-79e281072d88.png)

**Comment by vadimcn** [comment]
@x80819091: What does his profile picture have to do with anything?

**Comment by thernstig** [comment]
Is this really a VS Code "team decision", or is it just rob that has been hellbent on removing this since the start, to get his way through?

I cannot fathom how a much-belowed feature like this can get removed. It is way better than the not-so-good, to put it mildiy, GUI experience. VS Code is a tool for _**coders**_. The JSON editor is much more natural for any coder worth its salt.

**Comment by Rory-Sullivan** [comment]
Sadly it seems I am way too late to this discussion and the damage is already done. Nonetheless I feel I should weigh in to "pour one out" for a much loved feature. 🍻

**Comment by wenfangdu** [comment]
For those of you who downvoted this issue, please upvote the following reverse issue as well:

- https://github.com/microsoft/vscode/issues/129594

**Comment by lonix1** [comment]
I added feature request #130457 to port features from the old to the new editor. Please upvote. We need 20 upvotes in 60 days.

Then, according to their own rules, they can't ignore our pain. Of course it could go onto the backlog forever.

Nonetheless we need to make a formal request.

**Comment by Ticmea** [comment]
> I added feature request #130457 to port features from the old to the new editor. Please upvote. We need 20 upvotes in 60 days.

24 upvotes in 17 hours; well that was fast. Let's hope that this pace continues, so it becomes obvious just how much these features are missed.

**Comment by wenfangdu** [comment]
In case this is a permanent change, I proposed some features that mimic the split JSON editor when using the UI editor, please upvote the issue if you feel the mentioned features are useful:

- https://github.com/microsoft/vscode/issues/131390

**Comment by wenfangdu** [comment]
In case this is a permanent change, I proposed some features that mimic the split JSON editor when using the UI editor, please upvote the issue if you feel the mentioned features are useful:

- https://github.com/microsoft/vscode/issues/131390
